### PR TITLE
Include unpaid leave/ vacations/ sick pay in weekly messages

### LINKED ISSUE #35: Include unpaid leave/ vacations/ sick pay in weekly messages

### PR DESCRIPTION
**Pull request by DZotoff**
Fixed expected time calculations for different vacation types

### CODE REVIEW

**Review by Kseniia-Koshkina — APPROVED** [review]

File: src/meta-assistant/generic/time-utils.ts [hunk 1]
```diff
@@ -24,9 +24,11 @@ namespace TimeUtilities {
    *
    * @returns various date formats/ numbers for start and end of previous week
```
Comment (Jdallos) [inline src/meta-assistant/generic/time-utils.ts:26]: Missing @param docs

File: src/meta-assistant/generic/time-utils.ts [hunk 2]
```diff
@@ -61,6 +63,32 @@ namespace TimeUtilities {
...
+   */
+  export const checkIfVacationCaseExists = (
+    personId: number,
+    timeRegistrations: TimeRegistrations[],
+    nonProjectTimes: NonProjectTime[],
+    startDate: DateTime,
+    endDate: DateTime
+  ) => {  
+
+    let total_non_project_time = 0;
```
Comment (Jdallos) [inline src/meta-assistant/generic/time-utils.ts]: Why using snake case here, typically we use camel case throughout this code.
I see below we have snake case variables coming from the forecast API, ideally these should have been mapped to be renamed but it's not necessary now. But any variable we introduce should be camel case

File: src/meta-assistant/generic/time-utils.ts [hunk 3]
```diff
@@ -61,6 +63,32 @@ namespace TimeUtilities {
...
+    timeRegistrations.forEach((registration) => {
+      const { date, time_registered, non_project_time, person } = registration;
+      if (personId=== person && nonProjectTimes.map(nonProjectTime => nonProjectTime.id).includes(non_project_time)) {
```
Comment (Jdallos) [inline src/meta-assistant/generic/time-utils.ts]:
```suggestion
      if (personId === person && nonProjectTimes.map(nonProjectTime => nonProjectTime.id).includes(non_project_time)) {
```

Comment (Jdallos) [inline src/meta-assistant/generic/time-utils.ts]: Perhaps a comment here to explain what is happening could be really helpful, its not obvious to me how this check if a person was on vacation or not.

File: src/meta-assistant/slack/slack-utils.ts [hunk 4]
```diff
@@ -215,11 +216,11 @@ Have a great week!
 
     for (const userData of weeklyCombinedData) {
       const { slackId, personId, expected } = userData;
-
+      const vacation_time = TimeUtilities.checkIfVacationCaseExists(personId, timeRegistrations, nonProjectTimes, weekStartDate, weekEndDate);
```
Comment (Jdallos) [inline src/meta-assistant/slack/slack-utils.ts]: Why snake case? use camel case

File: src/tests/meta-assistant/__mocks__/forecastMocks.ts [hunk 5]
```diff
@@ -76,8 +76,109 @@ export const forecastMockTimeRegistrations: TimeRegistrations[] = [
...
+  {
+    id: 35051000,
+    person: 439678,
+    project: null,
+    non_project_time: 280370,
+    time_registered: 100,
+    date: "2024-03-19",
+    approval_status: "APPROVED"
+  }
+];
```
Comment (Jdallos) [inline src/tests/meta-assistant/__mocks__/forecastMocks.ts:152]:
```suggestion
];

```

File: src/meta-assistant/generic/time-utils.ts (hunk 3 above)
Comment (Jdallos) [reply src/meta-assistant/generic/time-utils.ts]: Is it the case that non project times are always time off?

File: src/tests/meta-assistant/__mocks__/forecastMocks.ts [hunk 6]
```diff
@@ -76,8 +76,109 @@ export const forecastMockTimeRegistrations: TimeRegistrations[] = [
...
+    is_internal_time: false
+  },
+  {
+    id: 280220,
+    name: "child sickness",
+    is_internal_time: false
+  },
+  {
+    id: 280335,
+    name: "payed off vacation",
```
Comment (Jdallos) [inline src/tests/meta-assistant/__mocks__/forecastMocks.ts]:
```suggestion
    name: "paid off vacation",
```

Comment (Jdallos) [reply src/tests/meta-assistant/__mocks__/forecastMocks.ts]: Is this an actual non_project_time string?

File: src/tests/meta-assistant/vacationCalculationTest.test.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+beforeEach(() => {
+  jest.clearAllMocks();
+});
+
+describe("Expected time left test, including different non project time types", () => {
+  const fakeStartDate = DateTime.fromISO("2024-03-18");
+  const fakeEndDate = DateTime.fromISO("2024-03-21");
+
+  describe.each(expectedResults)("Check vacation calculations", (fakePersonId, expectedResult) => {
```
Comment (Jdallos) [inline src/tests/meta-assistant/vacationCalculationTest.test.ts:17]: Do I understand correctly this is checking that the function returns the amount of non_project_time/ time off?

**Review by Jdallos — CHANGES_REQUESTED** [review]

File: src/tests/meta-assistant/__mocks__/forecastMocks.ts (hunk 6 above)
Comment (DZotoff) [reply src/tests/meta-assistant/__mocks__/forecastMocks.ts]: yes, it exists in forecast

File: src/meta-assistant/generic/time-utils.ts (hunk 3 above)
Comment (DZotoff) [reply src/meta-assistant/generic/time-utils.ts]: nonProjectTimes list was filtered to include only "internal: false" so we check if non_project_time of time registration is any of registered not internal, meaning not project related issues(Management, Internal communication (Slack, meetings etc.)).

File: src/tests/meta-assistant/vacationCalculationTest.test.ts (hunk 7 above)
Comment (DZotoff) [reply src/tests/meta-assistant/vacationCalculationTest.test.ts:17]: Yes, you are correct. However, this function not only checks the precise amount, but also checks if the calculations are correct.

**Review by Jdallos — APPROVED** [review]

**Pull request merged**